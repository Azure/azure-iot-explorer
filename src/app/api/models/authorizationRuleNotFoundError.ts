/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ERROR_TYPES } from './../../constants/apiConstants';

export class AuthorizationRuleNotFoundError extends Error {
    public requiredPermissions: string[];
    constructor(requiredPermissions: string[]) {
        super(ERROR_TYPES.AUTHORIZATION_RULE_NOT_FOUND);
        this.name = this.message;
        this.requiredPermissions = requiredPermissions;
    }
}
