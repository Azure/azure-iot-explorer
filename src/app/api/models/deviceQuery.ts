/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
export default interface DeviceQuery {
    deviceId: string;
    clauses: QueryClause[];
    continuationTokens: string[];
    currentPageIndex: number;
}

export interface QueryClause {
    parameterType?: ParameterType;
    operation?: OperationType;
    value?: string;
    isError?: boolean;
}

export enum ParameterType {
    // non pnp
    edge = 'capabilities.iotEdge',
    status = 'status',
    // pnp
    capabilityModelId = 'dcm',
    interfaceId = 'interface',
}

export enum OperationType {
    equals = '=',
    notEquals = '!=',
}

export enum DeviceCapability {
    edge = 'true',
    nonEdge = 'false'
}

export enum DeviceStatus {
    enabled = 'enabled',
    disabled = 'disabled'
}
