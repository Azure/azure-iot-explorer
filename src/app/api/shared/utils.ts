/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { createHmac } from 'crypto';
import { IoTHubConnectionSettings } from '../services/devicesService';
import { LIST_PLUG_AND_PLAY_DEVICES, SAS_EXPIRES_MINUTES } from '../../constants/devices';
import DeviceQuery, { QueryClause, ParameterType, OperationType } from '../models/deviceQuery';
import { RepoConnectionSettings } from '../services/digitalTwinsModelService';
import { AppEnvironment } from '../../constants/shared';

const MILLISECONDS_PER_SECOND = 1000;
const SECONDS_PER_MINUTE = 60;

export const enum PnPQueryPrefix {
    HAS_CAPABILITY_MODEL = 'HAS_CAPABILITYMODEL',
    HAS_INTERFACE = 'HAS_INTERFACE'
}

export interface GenerateSasTokenParameters {
    /**
     * URI prefix (by segment) of the endpoints that can be accessed with this token, starting with host name of the IoT hub (no protocol). For example, myHub.azure-devices.net/devices/device1
     */
    resourceUri: string;

    /**
     * The key to use for the token.
     */
    key: string;

    /**
     * Expiration time in minutes. Defaults to 5.
     */
    expiration?: number;

    /**
     * The name of the key being used. Omit if the token refers to device-registry credentials.
     */
    keyName?: string;
}

export const generateSasToken = (parameters: GenerateSasTokenParameters) => {
    let token = null;
    if (!!parameters.resourceUri && !!parameters.key) {
        const encodedUri = encodeURIComponent(parameters.resourceUri);

        const expires = Math.ceil((Date.now() / MILLISECONDS_PER_SECOND) + (parameters.expiration || SAS_EXPIRES_MINUTES) * SECONDS_PER_MINUTE);
        const toSign = encodedUri + '\n' + expires;

        const hmac = createHmac('sha256', new Buffer(parameters.key, 'base64'));
        hmac.update(toSign);
        const base64UriEncoded = encodeURIComponent(hmac.digest('base64'));

        token = `SharedAccessSignature sr=${encodedUri}&sig=${base64UriEncoded}&se=${expires}${parameters.keyName ? '&skn=' + parameters.keyName : ''}`;
    }
    return token;
};

export const generatePnpSasToken = (repositoryId: string, audience: string, secret: string, keyName: string) => {
    const now = new Date();
    const ms = 1000;
    const expiry = (now.setDate(now.getDate() + 1) / ms).toFixed(0);
    const encodedServiceEndpoint = encodeURIComponent(audience);
    const encodedRepoId = encodeURIComponent(repositoryId);
    const signature = [encodedRepoId, encodedServiceEndpoint, expiry].join('\n').toLowerCase();
    const sigUTF8 = new Buffer(signature, 'utf8');
    const secret64bit = new Buffer(secret, 'base64');
    const hmac = createHmac('sha256', secret64bit);
    hmac.update(sigUTF8);
    const hash = encodeURIComponent(hmac.digest('base64'));
    return `SharedAccessSignature sr=${encodedServiceEndpoint}&sig=${hash}&se=${expiry}&skn=${keyName}&rid=${repositoryId}`;
};

export const getRepoConnectionInfoFromConnectionString = (connectionString: string): RepoConnectionSettings => {
    const connectionObject: RepoConnectionSettings = {};
    connectionString.split(';')
    .forEach((segment: string) => {
        const keyValue = segment.split('=');
        switch (keyValue[0]) {
            case 'HostName':
            connectionObject.hostName = keyValue[1];
            break;
            case 'SharedAccessKeyName':
            connectionObject.sharedAccessKeyName = keyValue[1];
            break;
            case 'SharedAccessKey':
            connectionObject.sharedAccessKey = keyValue[1];
            break;
            case 'RepositoryId':
            connectionObject.repositoryId = keyValue[1];
            break;
            default:
            // we don't use other parts of connection string
            break;
        }
    });

    return connectionObject;
};

export const getConnectionInfoFromConnectionString = (connectionString: string): IoTHubConnectionSettings => {
    const connectionObject: IoTHubConnectionSettings = {};
    connectionString.split(';')
    .forEach((segment: string) => {
        const keyValue = segment.split('=');
        switch (keyValue[0]) {
            case 'HostName':
            connectionObject.hostName = keyValue[1];
            break;
            case 'SharedAccessKeyName':
            connectionObject.sharedAccessKeyName = keyValue[1];
            break;
            case 'SharedAccessKey':
            connectionObject.sharedAccessKey = keyValue[1];
            break;
            default:
            // we don't use other parts of connection string
            break;
        }
    });

    return connectionObject;
};

export const buildQueryString = (query: DeviceQuery) => {
    return query ?
        `${LIST_PLUG_AND_PLAY_DEVICES} ${queryToString(query)}` :
        LIST_PLUG_AND_PLAY_DEVICES;
};

export const queryToString = (query: DeviceQuery) => {
    if (query.deviceId) {
        return `WHERE STARTSWITH(devices.deviceId, '${query.deviceId}')`;
    }
    const clauseQuery = clauseListToString(query.clauses && query.clauses.filter(clause => !!clause.parameterType));
    if ('' !== clauseQuery) {
        return `WHERE (${clauseQuery })`;
    }
    return '';
};

export const clauseListToString = (clauses: QueryClause[]) => {
    return clauses && clauses.map(clause => clauseToString(clause)).join(' AND ') || '';
};

export const clauseToString = (clause: QueryClause) => {
    switch (clause.parameterType) {
        case ParameterType.capabilityModelId:
            return toPnPClause(PnPQueryPrefix.HAS_CAPABILITY_MODEL, clause.value);
        case ParameterType.interfaceId:
            return toPnPClause(PnPQueryPrefix.HAS_INTERFACE, clause.value);
        case ParameterType.edge:
            return toEdgeClause(clause.parameterType, clause.value);
        case ParameterType.status:
            return toDeviceStatusClause(clause.parameterType, clause.value);
        default:
            return clauseItemToString(clause.parameterType, clause.operation, clause.value);
    }
};

export const clauseItemToString = (fieldName: string, operation: OperationType, value: unknown) => {
    if (!fieldName || !operation || !value) {
        return;
    }
    return `${fieldName} ${operation} ${escapeValue(value as string)}`;
};

export const escapeValue = (value: string) => {
    const isString = new RegExp(/[^0-9.]/);
    const hasQuotes = new RegExp(/^'.*'$/);
    if (value.match(isString) && !value.match(hasQuotes)) {
        return `'${value}'`;
    }
    return value;
};

export const toPnPClause = (pnpFunctionName: string, value: string): string => {
    const urnVersionRegEx = new RegExp(/:[0-9]+$/);
    if (urnVersionRegEx.test(value)) {
        const urnVersion = urnVersionRegEx.exec(value)[0].replace(':', '');
        const urnName = value.replace(urnVersionRegEx, '');

        return `${pnpFunctionName}(${escapeValue(urnName)}, ${urnVersion})`;
    }
    return `${pnpFunctionName}('${value}')`; // when provided value is not urn with version, pass it in as string
};

export const toEdgeClause = (edgeFunctionName: string, value: string): string => {
    return `${edgeFunctionName}=${value === 'true' ? true : false}`;
};

export const toDeviceStatusClause = (edgeFunctionName: string, value: string): string => {
    return `${edgeFunctionName}='${value}'`;
};

export const isAppInElectron = () => {
    return process.env.NODE_ENV === AppEnvironment.DevelopmentElectron || process.env.NODE_ENV === AppEnvironment.ProdElectron;
};

export const isAppInDevelopmentMode = () => {
    return process.env.NODE_ENV === AppEnvironment.DevelopmentElectron || process.env.NODE_ENV === AppEnvironment.DevelopmentHosted;
};
