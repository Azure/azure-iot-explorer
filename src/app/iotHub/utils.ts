/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/

export const getShortHostName = (name: string) => {
    return name && name.replace(/\..*/, '');
};
