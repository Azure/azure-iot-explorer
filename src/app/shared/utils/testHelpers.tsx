/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { LocalizationContextProvider } from '../contexts/localizationContext';
import configureStore from '../../../app/shared/redux/store/configureStore';

export const testWithLocalizationContext = (Target: JSX.Element, enzymeWrapper: any = shallow) => { // tslint:disable-line:no-any
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children;

    return enzymeWrapper(<Children t={jest.fn((value: string) => value)}/>);
};

export const testWithLocalizationContextAndErrorBoundary = (Target: JSX.Element, enzymeWrapper: any = shallow) => { // tslint:disable-line:no-any
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children.props.children;

    return enzymeWrapper(<Children t={jest.fn()}/>);
};

export const testSnapshot = (Target: JSX.Element) => {
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children;
    expect(shallow(<Children t={jest.fn((value: string) => value)}/>)).toMatchSnapshot();
};

export const mountWithLocalization = (Target: JSX.Element, connectToStore: boolean = false, useMemoryRouter: boolean = false, routerInitialEntries = ['/']) => {
    let wrapper = (
        <LocalizationContextProvider value={{t: jest.fn((value: string) => value)}}>
            {Target}
        </LocalizationContextProvider>
    );

    if (connectToStore) {
        wrapper = (
            <Provider store={configureStore()}>
                {wrapper}
            </Provider>
        );
    }

    if ( useMemoryRouter ) {
        wrapper = (
            <MemoryRouter initialEntries={routerInitialEntries} keyLength={0}>
                {wrapper}
            </MemoryRouter>
        );
    }

    return mount(wrapper);
};
