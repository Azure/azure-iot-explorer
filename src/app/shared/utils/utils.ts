/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
export const generateKey = (generator: (byteArray: Uint8Array) => void = randomValueGenerator): string => {
    const byteLength = 32;
    const byteArray = new Uint8Array(byteLength);
    generator(byteArray);

    let bytes = '';
    for (let i = 0; i < byteLength; i++) {
        bytes += String.fromCharCode(byteArray[i]);
    }

    return btoa(bytes);
};

export const randomValueGenerator = (byteArray: Uint8Array) => {
    // tslint:disable-next-line:no-any
    const windowAny = window as any;  // msCrypto is not typed.
    const crypto = windowAny.crypto || windowAny.msCrypto;
    crypto.getRandomValues(byteArray);
};

export const validateKey = (key: string): boolean => {
    const pattern = new RegExp('^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$');
    return pattern.test(key);
};

export const validateThumbprint = (key: string): boolean => {
    const pattern = new RegExp('^[A-Fa-f0-9]{40}$');
    return pattern.test(key);
};

export const validateDeviceId = (key: string): boolean => {
    const pattern = new RegExp('^[A-Za-z0-9-:.+%_#*?!(),=@;$\']{0,128}$');
    return pattern.test(key);
};

export const validateModuleIdentityName = (key: string): boolean => {
    const pattern = new RegExp('^[A-Za-z0-9-:.+%_#*?!(),=@;$\']{0,128}$');
    return pattern.test(key);
};
