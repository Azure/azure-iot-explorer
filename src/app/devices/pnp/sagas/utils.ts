import { ModelIdCasingNotMatchingException } from '../../../shared/utils/exceptions/modelIdCasingNotMatchingException';
import { ModelDefinition } from '../../../api/models/modelDefinition';
import { ModelRepositoryConfiguration } from '../../../shared/modelRepository/state';
import { REPOSITORY_LOCATION_TYPE } from '../../../constants/repositoryLocationTypes';

export const getSplitInterfaceId = (fullName: string) => {
    // when component definition is inline, interfaceId is compose of parent file name and inline schema id concatenated with a slash
    return fullName.split('/');
};

export const getFlattenedModel = (model: ModelDefinition, splitInterfaceId: string[]) => {
    if (splitInterfaceId.length === 1) {
        return model;
    }
    else {
        // for inline component, the flattened model is defined under contents array's with matching schema @id
        const components = model.contents.filter((content: any) => // tslint:disable-line: no-any
            content['@type'] === 'Component' && typeof content.schema !== 'string' && content.schema['@id'] === splitInterfaceId[1]);
        return components[0];
    }
};

export const checkModelIdCasing = (model: ModelDefinition, id: string) => {
    if (model['@id'] !== id) {
        throw new ModelIdCasingNotMatchingException();
    }
};

export const getLocationSettingValue = (locations: ModelRepositoryConfiguration[], type: REPOSITORY_LOCATION_TYPE): string => {
    const filteredValue = locations.filter(location => location.repositoryLocationType === type)?.[0]?.value || '';
    return filteredValue.replace(/\/$/, ''); // remove trailing slash
};

export const getDmrParams = (path: string, interfaceId: string): {folderPath: string, fileName: string} => {
    // convert dtmi name to follow drm convention
    // for example: dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
    const fullPath = `${path.replace(/\/$/, '')}/${interfaceId.toLowerCase().replace(/:/g, '/').replace(';', '-')}.json`;
    // folderPath will be converted to for example: ${path}/dtmi/com/example, and fileName will be thermostat-1.json
    return {folderPath: fullPath.substring(0, fullPath.lastIndexOf('/')), fileName: fullPath.substring(fullPath.lastIndexOf('/') + 1, fullPath.length)};
};
