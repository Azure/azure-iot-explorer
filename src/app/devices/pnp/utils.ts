/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../shared/utils/queryStringHelper';
import { ROUTE_PARAMS } from '../../constants/routes';
import { getDeviceTwinAction, getModuleTwinAction } from './actions';
import { ModelDefinition } from '../../api/models/modelDefinition';
import { ComponentToModelId, JsonSchemaAdaptor } from '../../shared/utils/jsonSchemaAdaptor';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../constants/devices';

export const getBackUrl = (path: string, search: string) => {
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    let url = `${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`;
    if (moduleId) {
        url += `&${ROUTE_PARAMS.MODULE_ID}=${encodeURIComponent(moduleId)}`;
    }
    return url;
};

export const dispatchGetTwinAction = (search: string, dispatch: (action: any) => void) => {  // tslint:disable-line:no-any
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    if (moduleId) {
        dispatch(getModuleTwinAction.started({deviceId, moduleId}));
    }
    else {
        dispatch(getDeviceTwinAction.started(deviceId));
    }
};

export const getComponentNameAndInterfaceIdArray = (modelDefinition: ModelDefinition): ComponentToModelId[] => {
    if (!modelDefinition) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const components = jsonSchemaAdaptor.getComponentNameToModelIdMapping();
    // check if model contains no-component items
    if (jsonSchemaAdaptor.getNonWritableProperties().length +
        jsonSchemaAdaptor.getWritableProperties().length +
        jsonSchemaAdaptor.getCommands().length +
        jsonSchemaAdaptor.getTelemetry().length > 0) {
        components.unshift({
            componentName: DEFAULT_COMPONENT_FOR_DIGITAL_TWIN,
            modelId: modelDefinition['@id']
        });
    }
    return components;
};
