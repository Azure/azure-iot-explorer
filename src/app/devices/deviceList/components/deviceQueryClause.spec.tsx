/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import DeviceQueryClause from './deviceQueryClause';
import { testWithLocalizationContext } from '../../../shared/utils/testHelpers';
import { ParameterType, OperationType } from '../../../api/models/deviceQuery';

describe('components/devices/DeviceQueryClause', () => {
    // tslint:disable-next-line:no-any
    const event: any = {
        target: {
            title: 'name',
        }
    };

    it('matches snapshot', () => {
        const wrapper = testWithLocalizationContext(
            <DeviceQueryClause
                operation={OperationType.equals}
                parameterType={ParameterType.status}
                value="enabled"
                index={0}
                removeClause={jest.fn()}
                setClause={jest.fn()}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot without operation', () => {
        const wrapper = testWithLocalizationContext(
            <DeviceQueryClause
                parameterType={ParameterType.capabilityModelId}
                value="enabled"
                index={0}
                removeClause={jest.fn()}
                setClause={jest.fn()}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot without operation', () => {
        const wrapper = testWithLocalizationContext(
            <DeviceQueryClause
                parameterType={ParameterType.edge}
                value="true"
                index={0}
                removeClause={jest.fn()}
                setClause={jest.fn()}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('calls removeClause', () => {
        const removeClause = jest.fn();
        const wrapper = testWithLocalizationContext(
            <DeviceQueryClause
                parameterType={ParameterType.capabilityModelId}
                value="enabled"
                index={0}
                removeClause={removeClause}
                setClause={jest.fn()}
            />
        );
        const removeButton = wrapper.find(IconButton);
        removeButton.simulate('click');
        expect(removeClause).toBeCalled();
    });
});
