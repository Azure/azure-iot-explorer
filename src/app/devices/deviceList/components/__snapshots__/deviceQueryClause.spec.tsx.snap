// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/DeviceQueryClause matches snapshot 1`] = `
<section
  className="search-pill active no-operator"
  key="query-0"
>
  <StyledWithResponsiveMode
    ariaLabel="deviceLists.query.searchPills.clause.parameterType.ariaLabel"
    className="parameter-type"
    componentRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "capabilities.iotEdge",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.edge",
        },
        Object {
          "key": "status",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.status",
        },
        Object {
          "key": "dcm",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.capabilityModelId",
        },
        Object {
          "key": "interface",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.interfaceId",
        },
      ]
    }
    title="deviceLists.query.searchPills.clause.parameterType.title"
  />
  <StyledWithResponsiveMode
    ariaLabel="deviceLists.query.searchPills.clause.value.title"
    className="clause-value"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "enabled",
          "text": "deviceLists.query.searchPills.clause.value.deviceStatus.enabled",
        },
        Object {
          "key": "disabled",
          "text": "deviceLists.query.searchPills.clause.value.deviceStatus.disabled",
        },
      ]
    }
    selectedKey="enabled"
    title="deviceLists.query.searchPills.clause.value.title"
  />
  <CustomizedIconButton
    ariaLabel="deviceLists.query.searchPills.clause.remove.ariaLabel"
    className="remove-pill"
    data={0}
    iconProps={
      Object {
        "iconName": "Clear",
      }
    }
    onClick={[Function]}
    title="deviceLists.query.searchPills.clause.remove.title"
    type="button"
  />
</section>
`;

exports[`components/devices/DeviceQueryClause matches snapshot without operation 1`] = `
<section
  className="search-pill active no-operator"
  key="query-0"
>
  <StyledWithResponsiveMode
    ariaLabel="deviceLists.query.searchPills.clause.parameterType.ariaLabel"
    className="parameter-type"
    componentRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "capabilities.iotEdge",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.edge",
        },
        Object {
          "key": "status",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.status",
        },
        Object {
          "key": "dcm",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.capabilityModelId",
        },
        Object {
          "key": "interface",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.interfaceId",
        },
      ]
    }
    title="deviceLists.query.searchPills.clause.parameterType.title"
  />
  <StyledTextFieldBase
    ariaLabel="deviceLists.query.searchPills.clause.value.title"
    className="clause-value"
    onChange={[Function]}
    placeholder="deviceLists.query.searchPills.clause.value.placeholder"
    required={true}
    title="deviceLists.query.searchPills.clause.value.title"
  />
  <CustomizedIconButton
    ariaLabel="deviceLists.query.searchPills.clause.remove.ariaLabel"
    className="remove-pill"
    data={0}
    iconProps={
      Object {
        "iconName": "Clear",
      }
    }
    onClick={[Function]}
    title="deviceLists.query.searchPills.clause.remove.title"
    type="button"
  />
</section>
`;

exports[`components/devices/DeviceQueryClause matches snapshot without operation 2`] = `
<section
  className="search-pill active no-operator"
  key="query-0"
>
  <StyledWithResponsiveMode
    ariaLabel="deviceLists.query.searchPills.clause.parameterType.ariaLabel"
    className="parameter-type"
    componentRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "capabilities.iotEdge",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.edge",
        },
        Object {
          "key": "status",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.status",
        },
        Object {
          "key": "dcm",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.capabilityModelId",
        },
        Object {
          "key": "interface",
          "text": "deviceLists.query.searchPills.clause.parameterType.items.interfaceId",
        },
      ]
    }
    title="deviceLists.query.searchPills.clause.parameterType.title"
  />
  <StyledWithResponsiveMode
    ariaLabel="deviceLists.query.searchPills.clause.value.title"
    className="clause-value"
    onChange={[Function]}
    options={
      Array [
        Object {
          "ariaLabel": "deviceLists.query.searchPills.clause.value.deviceCapability.edge",
          "key": "true",
          "text": "deviceLists.query.searchPills.clause.value.deviceCapability.edge",
        },
        Object {
          "ariaLabel": "deviceLists.query.searchPills.clause.value.deviceCapability.nonEdge",
          "key": "false",
          "text": "deviceLists.query.searchPills.clause.value.deviceCapability.nonEdge",
        },
      ]
    }
    selectedKey="true"
    title="deviceLists.query.searchPills.clause.value.title"
  />
  <CustomizedIconButton
    ariaLabel="deviceLists.query.searchPills.clause.remove.ariaLabel"
    className="remove-pill"
    data={0}
    iconProps={
      Object {
        "iconName": "Clear",
      }
    }
    onClick={[Function]}
    title="deviceLists.query.searchPills.clause.remove.title"
    type="button"
  />
</section>
`;
